### ISSUE TITLE
Meson claims "Unknown compiler: javac" after it finds javac

### ISSUE THREAD

**Issue by langermarc19** [mesonbuild/meson#10900]
I am trying to compile the [rubberband](https://github.com/breakfastquay/rubberband) audio software library using meson. (as instructred from the rubberband readme) 

I get the following error: 
![Screen Shot 2022-10-06 at 5 02 25 PM](https://user-images.githubusercontent.com/29733599/194417972-6c086b21-fd6e-47e4-b1b6-5382f9b8bc70.png)

coming from these lines in the meson.build: 

```
if javac.found() and jar.found()
  jni_dep = dependency('jni', version: '>= 7.0.0', required: false)
  if not jni_dep.found()
    if cpp.has_header('jni.h', args: extra_include_args)
      jni_dep = declare_dependency()
    endif
  endif
else
  # Declare jni_dep so it isn't totally undefined - but it won't be
  # used below because we are conditional on javac/jar as well
  jni_dep = declare_dependency()
endif
```

I don't understand why if it's finding jar and javac it's then giving the unknown compiler error.

**Comment by eli-schwartz** [comment]
What is the output of `/usr/bin/javac -version`?

**Comment by eli-schwartz** [comment]
@tristan957 I realize that the dependency uses `compilers.detect_compiler_for()` which is the entry point for add_languages(), but elsewhere dependencies use `dependencies.base.detect_compiler()` that only searches if the project already added that language before. Any reason for this?

**Comment by tristan957** [comment]
No reason other than a misunderstanding it seems.

**Comment by langermarc19** [comment]
@eli-schwartz 

`marclanger@Marcs-Mac-mini ~ % /usr/bin/javac -version
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
`

**Comment by langermarc19** [comment]
I installed java and it worked. Thanks everyone.

**Comment by carlocab** [comment]
> Is this some sort of thing like where Windows has a built-in python program that just opens the Microsoft Store?

Kind of, but worse. `javac` is `/usr/bin/javac`, but Apple doesn't even provide a nice interface for installing a JDK. You have to provide one yourself and put it in the right place (`/Library/Java/JavaVirtualMachines`, last I checked).

**Comment by tristan957** [comment]
@eli-schwartz should a PR be submitted to swap the compiler detection out?

**Comment by eli-schwartz** [comment]
@tristan957 probably... at the very least, it's a bug if `dependency('jni', required: false)` aborts Meson with an error because it cannot find jni.

**Comment by eli-schwartz** [comment]
Rubberband in hg `tip` now works around this by checking for a Java language before looking up the dependency. Which is probably correct in the long term anyway. ;)

Please verify that it builds in homebrew, vcpkg, et. al.

**Comment by eli-schwartz** [comment]
This was also fixed in Meson via #10945
